src/Microsoft.Deployment.DotNet.Releases/src/ReleaseFile.cs

```diff
@@ -0,0 +1,143 @@
+{
+    /// <summary>
+    /// Represents a single file associated with a release component such as an SDK or runtime.
+    /// </summary>
+    public class ReleaseFile : IEquatable<ReleaseFile>
+    {
+        private static SHA512 HashAlgorithm = SHA512Managed.Create();
+
+        /// <summary>
+        /// The URL from where to download the file.
+        /// </summary>
+        public Uri Address
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// The filename and extension of this <see cref="ReleaseFile"/>.
+        /// </summary>
+        [JsonIgnore]
+        public string FileName => Path.GetFileName(Address.LocalPath);
+
+        /// <summary>
+        /// The <see cref="SHA512"/> hash of the file.
+        /// </summary>
+        public string Hash
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// The version agnostic name and extension of the file.
+        /// </summary>
+        public string Name
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// The runtime identifier associated with the file.
+        /// </summary>
+        public string Rid
+        {
+            get;
+            private set;
+        }
+
+        [JsonConstructor]
+        internal ReleaseFile([JsonProperty(PropertyName = "hash")] string hash,
+            [JsonProperty(PropertyName = "name")] string name,
+            [JsonProperty(PropertyName = "rid")] string rid,
+            [JsonProperty(PropertyName = "url")] string address)
+        {
+            Hash = hash;
+            Name = name;
+            Rid = rid;
+            Address = new Uri(address);
+        }
+
+        /// <summary>
+        /// Download this file to the specified local file and verify the file hash. If the hash is invalid, the
+        /// file will be deleted.
+        /// </summary>
+        /// <param name="destinationPath">The path, including the filename of the local file. The file will be
+        /// overwritten if it already exists.</param>
+        /// <exception cref="InvalidDataException">Thrown if the downloaded file's hash does to match the 
+        /// expected hash.</exception>
+        public async Task DownloadAsync(string destinationPath)
+        {
+            if (destinationPath is null)
+            {
+                throw new ArgumentNullException(nameof(destinationPath));
+            }
+
+            if (destinationPath == string.Empty)
+            {
+                throw new ArgumentException(string.Format(ReleasesResources.ValueCannotBeEmpty, nameof(destinationPath)));
+            }
+
+            await Utils.DownloadFileAsync(Address, destinationPath);
+
+            var actualHash = Utils.GetFileHash(destinationPath, HashAlgorithm);
+
+            if (!string.Equals(Hash, actualHash, StringComparison.OrdinalIgnoreCase))
+            {
+                File.Delete(destinationPath);
```

Comment:
As per our conversation, I've added an additional method that will download, but not verify and one that will download to a temp file, verify the hash and then either delete or move the temp file depending on the hash results.

I wasn't sure about adding a test because it could turn out to be flaky if there is a problem downloading the file at all